server/auth.go

```diff
@@ -78,15 +91,83 @@ func (h *Handler) LoginPost(c echo.Context) error {
 }
 
 func (h *Handler) LoginGet(c echo.Context) error {
+	sess, _ := session.Get(CookieKeySession, c)
+
 	vars := map[string]interface{}{
 		"csrf":               c.Get("csrf").(string),
 		"globus":             viper.GetBool("globus_signup"),
 		"enable_user_signup": viper.GetBool("enable_user_signup"),
+		"success":            sess.Flashes("success"),
 	}
 
+	sess.Save(c.Request(), c.Response())
+
 	return c.Render(http.StatusOK, "login.html", vars)
 }
 
+func (h *Handler) ChangePost(c echo.Context) error {
+	message := ""
+	sess, _ := session.Get(CookieKeySession, c)
+
+	expired := sess.Values["password-expired"]
+
+	if expired == nil || sess.Values["uid"] == nil || !expired.(bool) {
+		return echo.NewHTTPError(http.StatusUnauthorized, "Access denied.")
+	}
+
+	uid := sess.Values["uid"].(string)
+	current := c.FormValue("password")
+	pass := c.FormValue("new_password")
+	pass2 := c.FormValue("new_password2")
+	challenge := c.FormValue("challenge")
+
+	err := h.setPassword(uid, current, pass, pass2, challenge)
+	if err != nil {
+		message = err.Error()
+	} else {
+		sess.AddFlash("Password successfully changed!", "success")
+		delete(sess.Values, "password-expired")
+		delete(sess.Values, "uid")
+		sess.Save(c.Request(), c.Response())
+		return c.Redirect(http.StatusFound, Path("/auth/login"))
+	}
+
+	vars := map[string]interface{}{
+		"csrf":               c.Get("csrf").(string),
+		"globus":             viper.GetBool("globus_signup"),
+		"enable_user_signup": viper.GetBool("enable_user_signup"),
+		"message":            message,
+		"uid":                uid,
```

Comment:
We're missing `"otponly":            user.OTPOnly(),` in the POST request.

Reply:
For example, user has OTP enabled. The POST with an incorrect password, the template will be displayed again but it won't know they have OTP enabled.

Reply:
Ah, you're right. I'll add it too.

Reply:
Looks good. Thanks!